fix: Port 4000 is already occupied

chriskyfung opened the pull request:
This PR fixes #29 with the following changes:

- Stop Jekyll server before exiting VS Code within `deactivate()`.
- Enable user easily changes the port and the base URL configuration in the extension settings.
- Deprecated `openLocalJekyllSite()` in `open-in-browser.ts`
- Changed to get the server address (_including the **port** and **base URL**_) from stdout.
- Replaced `openLocalJekyllSite()` with `openUrl()`

chriskyfung committed:
fix: Port 4000 is already occupied (#29)

- Stop Jekyll server before exiting VS Code within deactivate().
- Enable user easily changes the port and the base URL configuration in the extension settings.
- Deprecated openLocalJekyllSite() in open-in-browser.ts
- Changed to get the server address (including the port and base URL) from stdout.
- Replaced openLocalJekyllSite() with openUrl()

Kanna727 commented on src/extension.ts:
```diff
@@ -452,4 +467,17 @@ export function activate(context: ExtensionContext) {
 
 // this method is called when your extension is deactivated
-export function deactivate() { }
+export function deactivate() {
```
I think the ability to stop Jekyll server before exiting VS Code should be an option to users which can be set in the extension settings with default being false. Let me know your thoughts

chriskyfung replied:
I think we can add an option in the extension settings but I prefer to stop Jekyll when exit VS Code. Otherwise, it is painful to restore the port and unable to monitor the Jekyll server in a new VS Code window.

I have no idea about why the Jekyll should continue to run in the background.

chriskyfung replied:
Can you suggest a name for the new extension setting? That I add it to the `deactivate()` and `package.json`.

chriskyfung replied:
I made a draft as belows:

```json
# package.json
"jekyll-run.stopServerOnExit": {
	"type": "boolean",
	"description": "Stop Jekyll server before exiting VS Code",
	"default": "false"
}
```

```ts
# extension.ts
export function deactivate() {
    const config = Config.get();
    if (config.stopServerOnExit) {
        console.log("Exit VSCode. Stopping: " + pid);
        commands.executeCommand('jekyll-run.Stop');
    }
}
```

Kanna727 replied:
Thanks @chriskyfung. Sometimes users might want just the extension to stop but not the server. So, let's provide this option for users to do whatever they want

Kanna727 commented on src/extension.ts:
```diff
@@ -452,4 +467,17 @@ export function activate(context: ExtensionContext) {
-export function deactivate() { }
+export function deactivate() {
+    console.log("Exit VSCode. Stopping: " + pid);
+    if (currWorkspace && isRunning) {
```
This code is duplicate of [Stop command](https://github.com/chriskyfung/jekyll-run/blob/master/src/extension.ts#L386). Please extract this into a separate function and use it at both the places

chriskyfung replied:
Make sense!

chriskyfung replied:
I changed to stop the server in `deactivate()` using the following code:

```ts
commands.executeCommand('jekyll-run.Stop');
```

I think it will be simpler and maintainable.

Kanna727 commented on src/extension.ts:
```diff
@@ -34,27 +40,36 @@ enum Icons {
 }
 
-function checkConfigAndGetBaseUrl(currWorkspace: WorkspaceFolder): boolean {
+function checkConfigAndGetPort(currWorkspace: WorkspaceFolder): boolean {
```
```suggestion
function getPortAndBaseurl(currWorkspace: WorkspaceFolder): boolean {
```

Kanna727 commented on src/extension.ts:
```diff
@@ -76,14 +91,12 @@ function isStaticWebsiteWorkspace(): boolean {
             currWorkspace = workspace.getWorkspaceFolder(resource);
             if (currWorkspace) {
-                checkConfigAndGetBaseUrl(currWorkspace);
                 return checkConfigAndGetPort(currWorkspace);
```
```suggestion
                return getPortAndBaseurl(currWorkspace);
```

Kanna727 commented on src/extension.ts:
```diff
@@ -76,14 +91,12 @@ function isStaticWebsiteWorkspace(): boolean {
         currWorkspace = workspace.workspaceFolders[0];
         if (currWorkspace) {
-            checkConfigAndGetBaseUrl(currWorkspace);
             return checkConfigAndGetPort(currWorkspace);
```
```suggestion
            return getPortAndBaseurl(currWorkspace);
```

Kanna727 commented on src/extension.ts:
```diff
@@ -34,27 +40,36 @@ enum Icons {
+    return { 
+        port: m_port ? true : false,
+        baseurl: m_baseurl ? true : false
+    };
```
```suggestion
```

Kanna727 commented on src/extension.ts:
```diff
@@ -34,27 +40,36 @@ enum Icons {
+
+        const configInArgs = getConfigFromArgs();
+        serverPort = configInArgs.port? portInArgs : portInConfig;
+        serverBaseurl = configInArgs.baseurl? baseurlInArgs : baseurlInConfig;
```
```suggestion
        getConfigFromArgs();
        serverPort = portInArgs || portInConfig;
        serverBaseurl = baseurlInArgs || baseurlInConfig;
```

Kanna727 commented on src/cmds/run.ts:
```diff
@@ -44,8 +45,12 @@ export class Run {
             outputChannel.show(true);
+            if (strString.includes('Server address')) {
+              const match = strString.match(/Server address: (\S+)/m);
+              this.address = match ? match[1] : `http://localhost:${serverPort}`;
```
```suggestion
              this.address = match?.[1]  || `http://localhost:${serverPort}`;
```

Kanna727 commented on src/extension.ts:
```diff
@@ -200,7 +213,7 @@ export function activate(context: ExtensionContext) {
     const open = commands.registerCommand('jekyll-run.Open', async () => {
         if (isStaticWebsiteWorkspace() && currWorkspace && isRunning) {
-            openLocalJekyllSite(portInConfig, baseurlInConfig);
+            openUrl(address);
```
Please change the default parameter value of [`openUrl`](https://github.com/chriskyfung/jekyll-run/blob/master/src/utils/open-in-browser.ts#L4) function to `http://127.0.0.1:4000`. Unfortunately, GitHub is not allowing me to comment on an unchanged line. Sorry for the inconvenience.

Kanna727 commented:
@chriskyfung Thanks a lot for the PR. Please complete the below tasks. Once done, I'll merge the PR.

- [ ] Do the review changes. Feel free to discuss.

chriskyfung committed:
Update src/extension.ts

Co-authored-by: Dedsec727  <chitturiprasanth13797@gmail.com>

chriskyfung committed:
Update src/extension.ts

Co-authored-by: Dedsec727  <chitturiprasanth13797@gmail.com>

chriskyfung committed:
Update src/cmds/run.ts

Co-authored-by: Dedsec727  <chitturiprasanth13797@gmail.com>

chriskyfung committed:
Update src/extension.ts

Co-authored-by: Dedsec727  <chitturiprasanth13797@gmail.com>

chriskyfung committed:
Update src/extension.ts

Co-authored-by: Dedsec727  <chitturiprasanth13797@gmail.com>

chriskyfung committed:
Update src/extension.ts

Co-authored-by: Dedsec727  <chitturiprasanth13797@gmail.com>

chriskyfung committed:
Changed http://127.0.0.1:4000 default URL to open

PR #47 Change the default parameter value of `openUrl` function to `http://127.0.0.1:4000`.

chriskyfung committed:
feat: set stop server on vscode exit in settings

PR #47 an option for users to set stop Jekyll server before exiting VS Code in the extension settings with default being false

chriskyfung committed:
fix: report error on running ruby code

Kanna727 commented:
Hi @chriskyfung, I forgot to mention but could you please submit another PR with updated README that explains the new config we've added?

Kanna727 commented:
:tada: This PR is included in version 1.7.0 :tada:

The release is available on [GitHub release](https://github.com/Kanna727/jekyll-run/releases/tag/v1.7.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
